src/schema/search.ts

```diff
@@ -0,0 +1,155 @@
+import {
+  GraphQLList,
+  GraphQLEnumType,
+  GraphQLString,
+  GraphQLNonNull,
+  GraphQLFieldConfig,
+} from "graphql"
+import { connectionFromArraySlice } from "graphql-relay"
+import { connectionWithCursorInfo } from "schema/fields/pagination"
+import { pageable } from "relay-cursor-paging"
+import { createPageCursors } from "schema/fields/pagination"
+import { SearchableItem } from "./searchableItem"
+
+import { Searchable } from "./searchable"
+import { convertConnectionArgsToGravityArgs } from "lib/helpers"
+
+import { visit, BREAK } from "graphql"
+
+export const SearchIndex = new GraphQLEnumType({
+  name: "SearchIndexes",
+  values: {
+    ARTWORK: {
+      value: "Artwork",
+    },
+    ARTIST: {
+      value: "Artist",
+    },
+  },
+})
+
+export const SearchMode = new GraphQLEnumType({
+  name: "SearchMode",
+  values: {
+    AUTOSUGGEST: {
+      value: "AUTOSUGGEST",
+    },
+    SITE: {
+      value: "SITE",
+    },
+  },
+})
+
+export const searchArgs = pageable({
+  query: {
+    type: new GraphQLNonNull(GraphQLString),
+    description: "Search query to perform. Required.",
+  },
+  indexes: {
+    type: new GraphQLList(SearchIndex),
+    description: "Indexes to include in search. Default: [ARTIST, ARTWORK].",
+  },
+  mode: {
+    type: SearchMode,
+    description: "Mode of search to exceute. Default: SITE.",
+  },
+})
+
+const hydrateSearchResult = (result, root) => {
+  const loaderMapping = {
+    Artist: root.artistLoader,
+    Artwork: root.artworkLoader,
+  }
+
+  const loader = loaderMapping[result.label]
+
+  return loader(result.id).then(object => {
+    return {
+      ...object,
+      __typename: result.label,
+    }
+  })
+}
+
+const shouldHydrate = info => {
+  let shouldHydrate = false
+
+  visit(info.fieldNodes[0], {
+    Field(node) {
+      if (node.name.value === "node") {
+        visit(node, {
+          InlineFragment(node) {
+            if (
+              node.typeCondition &&
+              node.typeCondition.name.value !== SearchableItem.name
+            ) {
+              shouldHydrate = true
+              return BREAK
+            }
+          },
+          FragmentSpread(_node) {
+            throw new Error(
+              "Named fragment spreads are currently unsupported for search."
+            )
+          },
+        })
+      }
+    },
+  })
+
+  return shouldHydrate
+}
+
+const SearchConnection = connectionWithCursorInfo(Searchable)
+
+export const Search: GraphQLFieldConfig<any, any, any> = {
+  type: SearchConnection,
+  description: "Global search",
+  args: searchArgs,
+  resolve: (source, args, _request, info) => {
+    const pageOptions = convertConnectionArgsToGravityArgs(args)
+
+    const gravityArgs = {
+      ...pageOptions,
+      indexes: args.indexes,
+      total_count: true,
+    }
+
+    const hydrate = shouldHydrate(info)
+
+    return source.searchLoader(gravityArgs).then(({ body, headers }) => {
+      const totalCount = parseInt(headers["x-total-count"])
```

Comment:
i dont think the `/match` and `/match/suggest` endpoints support returning counts. The count being placed in the header only happens 'for free' if you're using the conventional `paginate` helper in the API: https://github.com/artsy/gravity/blob/a9c7e6f229ad463bb5b8a00f38e0f65ebf1d6e78/app/api/util/page_helpers.rb#L30

Since the search endpoint returns an array of results, I don't think we can modify that payload, so we'll need to add the total into the header ourselves.

Basically, we need: `header 'X-Total-Count', results['hits']['total'].to_s if params[:total_count]` added in the endpoint: https://github.com/artsy/gravity/blob/6f831b886683f432d31a52cbcc893f68aac47943/app/api/v1/match_endpoint.rb#L195

Reply:
There was definitely a value ( 395 iirc) coming back, although it wasn’t immediately clear where in Gravity’s endpoint that was coming back from.

Reply:
yea, that's present in the main `/match` endpoint: https://github.com/artsy/gravity/blob/a9c7e6f229ad463bb5b8a00f38e0f65ebf1d6e78/app/api/v1/match_endpoint.rb#L178

aka `SITESEARCH`

but it's not present in the `/match/suggest` endpoint, which is where we need it for the new Autosuggest dropdown.